core/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -19,34 +19,44 @@
 
 package org.elasticsearch.tasks;
 
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.common.collect.Tuple;
 import org.elasticsearch.common.component.AbstractComponent;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
 import org.elasticsearch.common.util.concurrent.ConcurrentMapLong;
 import org.elasticsearch.transport.TransportRequest;
 
-import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.Iterator;
 import java.util.Map;
-import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.atomic.AtomicLong;
 
 /**
  * Task Manager service for keeping track of currently running tasks on the nodes
  */
-public class TaskManager extends AbstractComponent {
+public class TaskManager extends AbstractComponent implements ClusterStateListener{
 
     private final ConcurrentMapLong<Task> tasks = ConcurrentCollections.newConcurrentMapLongWithAggressiveConcurrency();
 
     private final AtomicLong taskIdGenerator = new AtomicLong();
 
+    private final Map<Tuple<String, Long>, String> banedParents = new ConcurrentHashMap<>();
+
     public TaskManager(Settings settings) {
         super(settings);
     }
 
+    public DiscoveryNodes lastDiscoveryNodes = null;
```

Comment:
can we make this private? I also wonder if we should initialize with `DiscoveryNodes#EMPTY_NODES` instead of null that would safe us a null check here and there?